src/library_pthread.js

```diff
@@ -416,6 +465,69 @@ var LibraryPThread = {
+    var transferedCanvasNames = {{{ makeGetValue('attr', 36, 'i32') }}}; // Comma-delimited list of IDs "canvas1, canvas2, ..."
+    if (transferedCanvasNames) {
+      transferedCanvasNames = Pointer_stringify(transferedCanvasNames);
+      transferedCanvasNames = transferedCanvasNames.split(',');
+    }
+    var offscreenCanvases = {}; // Dictionary of OffscreenCanvas objects we'll transfer to the created thread to own
+    var transferList = []; // List of JS objects that will transfer ownership to the Worker hosting the thread
+    for (var i in transferedCanvasNames) {
+      var name = transferedCanvasNames[i].trim();
+      var offscreenCanvas;
+      try {
+        if (name == '#canvas') {
+          if (GL.offscreenCanvases[Module['canvas'].id]) {
+            offscreenCanvas = GL.offscreenCanvases[Module['canvas'].id];
+            GL.offscreenCanvases[Module['canvas'].id] = null;
+          }
+          else if (Module['canvas'] instanceof OffscreenCanvas) {
+            offscreenCanvas = Module['canvas'];
+            Module['canvas'] = null;
+          }
+          else {
+            if (Module['canvas'].controlTransferredOffscreen) {
+              Module['printErr']('pthread_create: cannot transfer canvas with ID "' + name + '" to thread, since the current thread does not have control over it!');
+              return {{{ cDefine('EINVAL') }}};
+            }
+            offscreenCanvas = Module['canvas'].transferControlToOffscreen();
+            Module['canvas'].controlTransferredOffscreen = true;
+          }
+          offscreenCanvas.id = Module['canvas'].id;
+        } else if (GL.offscreenCanvases[name]) {
+          offscreenCanvas = GL.offscreenCanvases[name];
+          GL.offscreenCanvases[name] = null; // This thread no longer owns this canvas.
+        } else {
+          var canvas = document.getElementByID(name);
+          if (!canvas) {
+            Module['printErr']('pthread_create: could not find canvas with ID "' + name + '" to transfer to thread!');
+            return {{{ cDefine('EINVAL') }}};
+          }
+          if (canvas.controlTransferredOffscreen) {
+            Module['printErr']('pthread_create: cannot transfer canvas with ID "' + name + '" to thread, since the current thread does not have control over it!');
+            return {{{ cDefine('EINVAL') }}};
+          }
+          offscreenCanvas = canvas.transferControlToOffscreen();
+          canvas.controlTransferredOffscreen = true;
+          offscreenCanvas.id = canvas.id;
+        }
+        transferList.push(offscreenCanvas);
+        offscreenCanvases[offscreenCanvas.id] = offscreenCanvas;
+      } catch(e) {
+        Module['printErr']('pthread_create: failed to transfer control of canvas "' + name + '" to OffscreenCanvas! Error: ' + e);
+        return {{{ cDefine('EINVAL') }}};
+      }
+    }
+
+    // Synchronously proxy the thread creation to main thread if possible. If we need to transfer ownership of objects, then
```

Comment:
The sync version is faster, since it can be served while the main thread is still executing a cancellation point, whereas the async version must wait until the browser main thread yields execution back to browser.